sympy/matrices/expressions/hadamard.py

```diff
@@ -67,6 +66,9 @@ class HadamardProduct(MatrixExpr):
 
     def __new__(cls, *args, evaluate=False, check=None):
         args = list(map(sympify, args))
+        if len(args) == 0:
+            # We currently don't have a way to support one-matrices of generic dimensions:
+            raise ValueError("HadamardProduct needs at least one argument")
```

Comment:
We really should generalize GenericMatrix. I think the approach for handling unknown/generic shapes needs to be done much better in the matrix expressions.

Reply:
That is interesting. Though I would improve the assumptions system first (is there a way to represent expressions with conditions?).

Reply:
Do we have a way to create an expression with assumptions? We have the `with assuming(...): ... ` form in the assumptions module, but that is not a SymPy tree-expression.

What about something similar to Wolfram Mathematica's [Assuming](https://reference.wolfram.com/language/ref/Assuming.html)?

Reply:
There is no way to create an expression with assumptions.

Reply:
`assuming` is supposed to be the equivalent of `Assuming`. However, in general, the design of the new assumptions is that assumptions are separate from expressions. 

It's really still an open question of how to most cleanly handle this in the API, e.g., should every function pass assumptions around, or should we try to store the assumptions on expressions, or store them globally, or something else.

Reply:
Opened https://github.com/sympy/sympy/issues/24149 for this.